File: src/test/java/cube8540/oauth/authentication/oauth/token/domain/OAuth2AuthorizedRefreshTokenTest.java
package cube8540.oauth.authentication.oauth.token.domain;

import cube8540.oauth.authentication.AuthenticationApplication;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.Clock;

import static cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenTestHelper.REFRESH_EXPIRATION_DATETIME;
import static cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenTestHelper.makeAccessToken;
import static cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenTestHelper.makeRefreshTokenIdGenerator;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DisplayName("리플레시 토큰 도메인 테스트")
class OAuth2AuthorizedRefreshTokenTest {

    @Test
    @DisplayName("현재 시간이 만료일을 초과 했을시")
    void whenCurrentTimeExceedsExpirationDate() {
        OAuth2AuthorizedRefreshToken refreshToken = new OAuth2AuthorizedRefreshToken(makeRefreshTokenIdGenerator(), REFRESH_EXPIRATION_DATETIME, makeAccessToken());

        Clock clock = Clock.fixed(REFRESH_EXPIRATION_DATETIME.plusNanos(1).toInstant(AuthenticationApplication.DEFAULT_ZONE_OFFSET), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId());
        OAuth2AuthorizedRefreshToken.setClock(clock);

        assertTrue(refreshToken.isExpired());
        assertEquals(0, refreshToken.expiresIn());
    }

    @Test
    @DisplayName("현재 시간이 만료일을 초과 하지 않을시")
    void whenCurrentTimeNotExceedsExpirationDate() {
        OAuth2AuthorizedRefreshToken refreshToken = new OAuth2AuthorizedRefreshToken(makeRefreshTokenIdGenerator(), REFRESH_EXPIRATION_DATETIME, makeAccessToken());

        long expiresIn = 10;
        Clock clock = Clock.fixed(REFRESH_EXPIRATION_DATETIME.minusSeconds(expiresIn).toInstant(AuthenticationApplication.DEFAULT_ZONE_OFFSET), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId());
        OAuth2AuthorizedRefreshToken.setClock(clock);

        assertFalse(refreshToken.isExpired());
        assertEquals(expiresIn, refreshToken.expiresIn());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/token/domain/OAuth2AuthorizedRefreshTokenTest.kt
package cube8540.oauth.authentication.oauth.token.domain

import cube8540.oauth.authentication.AuthenticationApplication
import cube8540.oauth.authentication.toDefaultInstance
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.time.Clock
import java.time.LocalDateTime

class OAuth2AuthorizedRefreshTokenTest {

    private val tokenIdGenerator: OAuth2TokenIdGenerator = mockk {
        every { generateTokenValue() } returns OAuth2TokenId("tokenId")
    }
    private val refreshExpirationDateTime = LocalDateTime.of(2020, 1, 29, 11, 9)
    private val refreshToken: OAuth2AuthorizedRefreshToken = OAuth2AuthorizedRefreshToken(tokenIdGenerator, refreshExpirationDateTime, mockk(relaxed = true))

    @Nested
    inner class ExpirationTest {

        @Test
        fun `refresh token is expiration`() {
            val datetime = refreshExpirationDateTime.plusNanos(1)

            OAuth2AuthorizedRefreshToken.clock = Clock.fixed(datetime.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())

            val isExpired = refreshToken.isExpired()
            val expires = refreshToken.expiresIn()
            assertThat(isExpired).isTrue
            assertThat(expires).isEqualTo(0)
        }

        @Test
        fun `refresh token is not expiration`() {
            val datetime = refreshExpirationDateTime.minusSeconds(10)

            OAuth2AuthorizedRefreshToken.clock = Clock.fixed(datetime.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())

            val isExpired = refreshToken.isExpired()
            val expires = refreshToken.expiresIn()
            assertThat(isExpired).isFalse
            assertThat(expires).isEqualTo(10)
        }
    }
}